### PR TITLE
Add more elements until the container is filled

### PR DESCRIPTION
**Pull request by rleite**
If we can't fill the entire container on the first call the component will not fully filled until there is user interaction.
This is problematic if we have a small number of elements or if screen is really big.

This implementation will check if the container is fully filled after the digest of the handler call.

### CODE REVIEW

File: src/infinite-scroll.js [hunk 1]
```diff
@@ -131,7 +129,7 @@ angular.module(MODULE_NAME, [])
         return throttled;
       }
 
-      const handler = (THROTTLE_MILLISECONDS != null) ?
```
Comment (graingert) [inline src/infinite-scroll.js]: is handler reassigned?

Comment (rleite) [inline src/infinite-scroll.js]: true., I'll change back to const declaration

File: test/spec/ng-infinite-scroll.spec.js [hunk 2]
```diff
@@ -37,7 +47,7 @@ const getTemplate = (angularVersion, container, attrs, throttle) =>
       .run(function ($rootScope) {
         $rootScope.items = [];
         $rootScope.loadMore = function () {
-          [].push.apply($rootScope.items, new Array(100));
+          [].push.apply($rootScope.items, new Array(${nElements}));
```
Comment (graingert) [inline test/spec/ng-infinite-scroll.spec.js]: Templated code sucks, load this in with a data attribute on head with JSON in.

Comment (rleite) [inline test/spec/ng-infinite-scroll.spec.js]: I'm trying to be consistent with the rest of the template. Care to explain your suggestion?

Comment (graingert) [inline test/spec/ng-infinite-scroll.spec.js]:
```js
const data = { nElements };
return `<!doctype html>
  <head data-infinite-scroll-test=${escape(data)>
...
   var data = document.getElementByTagName('html').dataset;
...
[].push.apply($rootScope.items, new Array(data.nElements));
`;
```

Comment (rleite) [inline test/spec/ng-infinite-scroll.spec.js]: 👍

File: src/infinite-scroll.js (hunk 1 above)
Comment (rleite) [inline src/infinite-scroll.js]: Actually it needs to be let as it is resigned. es6lint doesn't allow for declaration before use.

File: src/infinite-scroll.js [hunk 3]
```diff
@@ -80,21 +74,25 @@ angular.module(MODULE_NAME, [])
...
+        if (checkWhenEnabled && scrollEnabled) {
+          if (scope.$$phase || $rootScope.$$phase) {
+            scope.infiniteScroll();
+          } else {
+            scope.$apply(scope.infiniteScroll);
           }
-        } else {
-          if (checkInterval) { $interval.cancel(checkInterval); }
-          checkWhenEnabled = false;
+          $timeout(handler);
```
Comment (graingert) [inline src/infinite-scroll.js]: does this need to be recursive?

Comment (rleite) [inline src/infinite-scroll.js]: not fully recursive as we have timeout and throttle in between but the idea is to check again after the digest is done.
Digest should make the dom change son that digest or trigger scrollEnabled = false if it is async calls.

File: src/infinite-scroll.js (hunk 1 above)
Comment (graingert) [inline src/infinite-scroll.js]: move the should throttle or not logic into the throttle function, then we can just do:

const handler = throttle(unthrottledHandler, THROTTLE_MILLISECONDS)

File: src/infinite-scroll.js (hunk 3 above)
Comment (graingert) [inline src/infinite-scroll.js]: that's fully recursive....

Comment (rleite) [inline src/infinite-scroll.js]: Yes but the point here is that I do think we need a this call here

**Comment by graingert** [comment]
You don't know how long the infiniteScroll callback will take and a timeout poll is not really acceptable.

This should only work if the infiniteScroll function returns a promise.

**Comment by rleite** [comment]
I don't think it matters since disabled should be used to synchronize with async call right?
If this is a synchronize call digest takes care of the work first.

**Comment by graingert** [comment]
@rleite ?

File: src/infinite-scroll.js (hunk 1 above)
Comment (rleite) [inline src/infinite-scroll.js]: The problem is that defaultHandler calls handler.

Comment (graingert) [inline src/infinite-scroll.js]: sure, what should actually happen is something like:

```
async function doScroll() {
    ...
    while(shouldScroll() && scrollEnabled) {
        result = scope.infiniteScroll();
        if (typeof result.then === 'function') {
            await scope.infiniteScroll();
            await $timeout(THROTTLE_MILLISECONDS);
        } else {
            await $timeout(THROTTLE_MILLISECONDS);
            return;
        }
    }
    ...
}
function getHandler() {
     let prom = null;
     async function resetProm(doScrollProm) {
         await doScrollProm;
         prom = null;
     }
     async function innerHandler() {
        if (!prom) {
             prom = resetProm(doScroll());
        }
        await prom;
    }
    return innerHandler;
}

const handler = getHandler();
```

The idea, throttling is all handled by doScroll, every time it's called it loops until the container is full, while handling the throttling. However if infiniteScroll is a regular funciton with no promise, doScroll just defaults to the previous behaviour.

How to convert this from ES2017 async/await into $q then chains and support all the other features and checks I've left out is left as an exercise for the reader.

**Pull request closed without merging**